File: GitToolBox/src/main/java/zielu/gittoolbox/util/GtStringUtil.java
package zielu.gittoolbox.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class GtStringUtil {
  private GtStringUtil() {
    throw new IllegalStateException();
  }

  @Nullable
  public static String firstLine(@Nullable String input) {
    if (input == null) {
      return null;
    } else {
      return findFirstLine(input);
    }
  }

  @NotNull
  private static String findFirstLine(@NotNull String input) {
    for (int i = 0, n = input.length(); i < n; i++) {
      int codePointAt = input.codePointAt(i);
      if (codePointAt == '\n' || codePointAt == '\r') {
        return input.substring(0, i);
      }
    }
    return input;
  }
}

File: src/main/kotlin/zielu/gittoolbox/util/GtStringUtil.kt
package zielu.gittoolbox.util

internal object GtStringUtil {
  @JvmStatic
  fun firstLine(input: String?): String? {
    return input?.let { findFirstLine(it) }
  }

  private fun findFirstLine(input: String): String {
    var i = 0
    val n = input.length
    while (i < n) {
      val codePointAt = input.codePointAt(i)
      if (codePointAt == '\n'.toInt() || codePointAt == '\r'.toInt()) {
        return input.substring(0, i)
      }
      i++
    }
    return input
  }
}
